components/raftstore/src/store/fsm/peer.rs

```diff
@@ -4719,6 +4748,25 @@ where
         let leader_id = self.fsm.peer.leader_id();
         let request = msg.get_requests();
 
+        // peer_id must be the same as peer's.
+        if let Err(e) = util::check_peer_id(msg, self.fsm.peer.peer_id()) {
+            self.ctx
+                .raft_metrics
+                .invalid_proposal
+                .mismatch_peer_id
+                .inc();
+            return Err(e);
+        }
+
+        // Forbid reads and writes when it's a witness except transfer leader
```

Comment:
Technically speaking, witness can write. Though not supporting read is useless when txn kv is used.

Reply:
Yes, and stop write can accelerate the transfer leader speed